src/docs/binding-generics.md

```diff
@@ -236,3 +236,194 @@ The proposed API will only work with types known at compile time. In some cases
+        self.first = newFirst
+    }
+}
+```
+
+The init method will behave just like a regular generic function returning generic type (see `ReturnData`), that means that we need to pass `SwiftIndirectResult`, pointer to data and metadata pointer.
+
+The `changeFirst` method will require additional handling. Upon inspection we can see that when calling instance methods on generic types Swift expects to be handled type metadata for the struct (in addition to `newFirst` pointer and `SwiftSelf`).
+
+Accounting for those two the projection could look something like:
+
+```csharp
+class Pair<T, U> : ISwiftObject
+{
+
+    private static nuint PayloadSize =  /* Extract size */;
+    private NativeHandle _payload;
+
+    public unsafe Pair(T first, U second)
+    {
+        _payload = new NativeHandle(NativeMemory.Alloc(PayloadSize));
+
+        SwiftIndirectResult swiftIndirectResult = new SwiftIndirectResult(_payload);
+
+        TypeMetadata.TryGetTypeMetadata<T>(out var firstMetadata);
+        TypeMetadata.TryGetTypeMetadata<U>(out var secondMetadata);
+
+        var nativeHandleFirst = Runtime.GetPayload(ref first);
+        var nativeHandleSecond = Runtime.GetPayload(ref second);
+
+        PairPInvokes.Pair(swiftIndirectResult, nativeHandleFirst, nativeHandleSecond, firstMetadata!.Value, secondMetadata!.Value);
+    }
+
+    public static TypeMetadata Metadata
+    {
+        /* This should be cached */
+        get
+        {
+            TypeMetadata.TryGetTypeMetadata<T>(out var firstMetadata);
+            TypeMetadata.TryGetTypeMetadata<U>(out var secondMetadata);
+
+            return PairPInvokes.PInvokeMetadata(TypeMetadataRequest.Complete, firstMetadata!.Value, secondMetadata!.Value);
+        }
+    }
+
+    public static object FromPayload(NativeHandle payload)
+    {
+        return new Pair<T, U>(payload);
+    }
+
+    public NativeHandle GetPayload() => _payload;
+
+    public unsafe void ChangeFirst(T newFirst)
+    {
+        SwiftSelf self = new SwiftSelf(_payload);
+        PairPInvokes.ChangeFirst(Runtime.GetPayload(ref newFirst), Metadata, self);
+    }
+
+    private unsafe Pair(NativeHandle payload)
+    {
+        _payload = new NativeHandle(NativeMemory.Alloc(PayloadSize));
+        NativeMemory.Copy(payload, _payload, PayloadSize);
+    }
+}
+```
+
+#### Frozen and non-frozen generic structs
+
+The frozen keyword can be applied to a generic struct just like it can be applied to a regular struct. However because the generic parameters type (and hence size and layout) is unknown at the compile time, so is the size and the layout of the generic struct. Specifically, the generic parameter can be a frozen or a non-frozen struct.
+
+If we consider the four cases:
+
+1. Non-frozen struct and at least one non-frozen generic parameter
+2. Non-frozen struct and (recursively) frozen generic parameters
+3. Frozen struct and at least one non-frozen generic parameter
+4. Frozen struct and (recursively) frozen generic parameters
```

Comment:
Like the way you're breaking this out into the possible cases.

Reply:
I cannot speak about whether it is recommended. It is definitely supported. There are some frozen generics in `SwiftUI`. I do not think it is possible to enforce using frozen parameters (maybe I am wrong), so it only up to the user of such generic struct to be kind enough to instantiate it with frozen types.

Reply:
You're right, more info here https://github.com/swiftlang/swift-evolution/blob/main/proposals/0260-library-evolution.md

> It does not necessarily have a known size or alignment. A generic frozen struct's layout might depend on the generic argument provided at run time.

> Even concrete instantiations may not have a known size or alignment. A frozen struct with a field that's a *non-*frozen, has a size that may not be known until run time.